Update DoctrineHelper

p365labs opened the pull request:
remove DoctrineHelper as it's a duplicate of Doctrine Common ClassUtils, 

Have a look a [this discussion](https://github.com/doctrine/common/issues/826#issuecomment-472456562) and this is the class which will substitute [the Common ClassUtils]
(https://github.com/Ocramius/ProxyManager/blob/addb7d39c990d3d2299315f0200e78ebe658cd7a/src/ProxyManager/Inflector/ClassNameInflector.php#L41)

maxhelias commented:
It's not a duplicate of `ClassUtils` because this one is deprecated. So we this change, we introduce a deprecated log

maxhelias commented:
But we can improve this behavior with this implementation : https://github.com/api-platform/core/blob/master/src/Util/ClassInfoTrait.php#L38

p365labs commented:
I said duplicate because the code inside DoctrineHelper is exactly the same the one is inside *Doctrine-> ClassUtils*, so if it's deprecated for Doctrine .. it's deprecated also for this project.

In my implementation, I simply remove DoctrineHelper and used the deprecated solution, which is 
not forbidden, but it's a statement to let us know we need to find a different strategy.

*Ocramius* suggested the right replacement strategy:
https://github.com/doctrine/common/issues/826#issuecomment-472456562
which is [this class](https://github.com/Ocramius/ProxyManager/blob/2.5.0/src/ProxyManager/Inflector/ClassNameInflector.php)

Did not know about the API-Platform solution. But why using that solution instead of the ProxyManager one?

maxhelias commented:
That correct for the duplicate but not for the deprecated for the project, it allows us to detach ourselves from doctrine depreciation and to be able to have compatibility with several versions without deprecated.

The replacement suggested by Ocramius, it's a good alternative but adds a hard dependency. So the good way would be to read the metadata with the EntityManager like this https://github.com/doctrine/common/issues/867 : 
```
$em = $this->getDoctrineEntityManager();
$classname = $em->getClassMetadata(get_class($object))->getName();
```
But i seems there is a problem with doctrine odm, so we need to integration like Api-Platform to be compatible with both proxy markers and without hard dependency : 
```
__CG__: Doctrine Common Marker for Proxy (ODM < 2.0 and ORM < 3.0)
__PM__: Ocramius Proxy Manager (ODM >= 2.0)
```

p365labs commented:
thnk you @DamienHarper & @maxhelias thank u for the clarification, yep for sure! I'll do it in the next couple of days.

p365labs commented:
Hi! I've updated the PR implementing the solution from ApiPlatform and referencing the authors.
Instead of using a trait I implemented a static method, and I think before merging we need more tests on this use case, I'll work on this during these days.

p365labs commented:
I'm wondering why inside DoctrineHelper we need a method ``generateProxyClassName``
is it a mandatory requirements?
At the moment is not used inside the ``src`` and it's only tested in ``tests``

Otherwise I need to find a way to identify which ORM or ODM version is using...which is a bit weird to do. 
Any ideas?

DamienHarper commented:
@p365labs I'm afraid it's a leftover from the past :) You can remove it from `DoctrineHelper` and from relevant tests

p365labs commented:
tests done!

DamienHarper commented on tests/DoctrineAuditBundle/Helper/DoctrineHelperTest.php:
```diff
@@ -11,18 +12,31 @@
+        self::assertSame(Comment::class, DoctrineHelper::getRealClassName($comment));
+
+        // __CG__: Doctrine Common Marker for Proxy (ODM < 2.0 and ORM < 3.0)
+        self::assertSame(Post::class, DoctrineHelper::getRealClassName('Proxies\__CG__\DH\DoctrineAuditBundle\Tests\Fixtures\Core\Standard\Post'));
```
In fact the `DoctrineHelper::generateProxyClassName()` method was only there to help not to write such FQCN.
BTW, as it wasn't used elsewhere, it's ok to have it removed.
